table/block_based_table_reader.cc

```diff
@@ -2491,18 +2493,28 @@ void BlockBasedTableIterator<TBlockIter, TValue>::InitDataBlock() {
 
 template <class TBlockIter, typename TValue>
 void BlockBasedTableIterator<TBlockIter, TValue>::FindKeyForward() {
-  assert(!is_out_of_bound_);
   // TODO the while loop inherits from two-level-iterator. We don't know
   // whether a block can be empty so it can be replaced by an "if".
   while (!block_iter_.Valid()) {
     if (!block_iter_.status().ok()) {
       return;
     }
+    bool block_is_out_of_bound = false;
+    if (read_options_.iterate_upper_bound != nullptr &&
+        block_iter_points_to_real_block_) {
+      // Do not set is_out_of_bound_ here. Even when index key is out of bound,
+      // it can be larger than smallest key of the next file on the level, and
+      // that file may still be within bound.
+      block_is_out_of_bound =
+          (user_comparator_.Compare(*read_options_.iterate_upper_bound,
+                                    index_iter_->user_key()) <= 0);
+    }
     ResetDataIter();
-    // We used to check the current index key for upperbound.
-    // It will only save a data reading for a small percentage of use cases,
-    // so for code simplicity, we removed it. We can add it back if there is a
-    // significnat performance regression.
+    if (block_is_out_of_bound) {
+      // The next block is out of bound. No need to read it.
+      TEST_SYNC_POINT_CALLBACK("BlockBasedTableIterator:out_of_bound", nullptr);
+      return;
```

Comment:
Let me try to understand correctly. In this case, we invalidate the iterator, but don't set upper bound flag. Does it mean that this flag will be false https://github.com/facebook/rocksdb/blob/v5.18.3/db/version_set.cc#L648 and always we are going to move to the next file?

Reply:
Yes, `is_out_of_bound_` flag will be false in this case. But `LevelIterator::KeyReachedUpperBound` will catch it and avoid reading the next file.

Reply:
Alternatively, if the block is not the last block, it is safe to set `is_out_of_bound_=true`. But we need to advance index iterator to know that, which risk doing extra IO (in case of partitioned-index).

Reply:
I think advancing index iterator is a better idea. I don't like so much cross layer contracts.

Reply:
done.